packages/vm/src/memory.rs

```diff
@@ -49,33 +53,34 @@ pub fn read_region(ctx: &Ctx, ptr: u32) -> Vec<u8> {
 /// A prepared and sufficiently large memory Region is expected at ptr that points to pre-allocated memory.
 ///
 /// Returns number of bytes written on success.
-pub fn write_region(ctx: &Ctx, ptr: u32, data: &[u8]) -> Result<usize, Error> {
-    let region = to_region(ctx, ptr);
-    let region_size = region.len as usize;
+pub fn write_region(ctx: &Ctx, ptr: u32, data: &[u8]) -> Result<(), Error> {
+    let mut region = to_region(ctx, ptr);
 
-    if data.len() > region_size {
+    let region_capacity = region.capacity as usize;
+    if data.len() > region_capacity {
         return RegionTooSmallErr {
-            size: region_size,
+            size: region_capacity,
             required: data.len(),
         }
         .fail();
     }
 
     // A performance optimization
     if data.is_empty() {
-        return Ok(0);
+        return Ok(());
     }
 
     let memory = ctx.memory(0);
 
-    match WasmPtr::<u8, Array>::new(region.offset).deref(memory, 0, region.len) {
+    match WasmPtr::<u8, Array>::new(region.offset).deref(memory, 0, region.capacity) {
         Some(cells) => {
             // In case you want to do some premature optimization, this shows how to cast a `&'mut [Cell<u8>]` to `&mut [u8]`:
             // https://github.com/wasmerio/wasmer/blob/0.13.1/lib/wasi/src/syscalls/mod.rs#L79-L81
             for i in 0..data.len() {
                 cells[i].set(data[i])
             }
-            Ok(data.len())
+            region.length = data.len() as u32;
```

Comment:
This is key, but here you just set it in the Region copy in Rust memory.
We need to update the Region object in wasm. 
Some analouge to `to_region`, like `update_wasm_region`, which updates the data at the wasm memory location

Reply:
Well spotted, thanks. That's a bug for sure.

Reply:
Implementted in https://github.com/CosmWasm/cosmwasm/pull/187/commits/0e4c72968174817254ae9a912c77b513167825a2